odxtools/diaglayers/protocol.py

```diff
@@ -0,0 +1,62 @@
+# SPDX-License-Identifier: MIT
+from copy import deepcopy
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, cast
+from xml.etree import ElementTree
+
+from ..comparamspec import ComparamSpec
+from ..exceptions import odxassert
+from ..odxlink import OdxDocFragment
+from ..protstack import ProtStack
+from .hierarchyelement import HierarchyElement
+from .protocolraw import ProtocolRaw
+
+
+@dataclass
+class Protocol(HierarchyElement):
+    """This is the class for primitives that are common for a given communication protocol
+
+    Most importantly this diagnostic layer is responsible for defining
+    the communication parameters that ought to be used.
+    """
+
+    @property
+    def protocol_raw(self) -> ProtocolRaw:
+        return cast(ProtocolRaw, self.diag_layer_raw)
+
+    @property
+    def comparam_spec(self) -> ComparamSpec:
+        return self.protocol_raw.comparam_spec
+
+    @property
+    def prot_stack(self) -> Optional[ProtStack]:
+        return self.protocol_raw.prot_stack
+
+    @staticmethod
+    def from_et(et_element: ElementTree.Element, doc_frags: List[OdxDocFragment]) -> "Protocol":
+        protocol_raw = ProtocolRaw.from_et(et_element, doc_frags)
+
+        return Protocol(diag_layer_raw=protocol_raw)
+
+    def __post_init__(self) -> None:
+        odxassert(
+            isinstance(self.diag_layer_raw, ProtocolRaw),
+            "The raw diagnostic layer passed to Protocol "
+            "must be a ProtocolRaw")
+
+    def __deepcopy__(self, memo: Dict[int, Any]) -> Any:
+        """Create a deep copy of the protocol layer
+
+        Note that the copied diagnostic layer is not fully
+        initialized, so `_finalize_init()` should to be called on it
+        before it can be used normally.
+        """
+
+        result = super().__deepcopy__(memo)
+
+        # note that the self.protocol_raw object is *not* copied at
+        # this place because the attribute points to the same object
+        # as self.diag_layer_raw.
+        result.protocol_raw = deepcopy(self.protocol_raw, memo)
+
+        return result
```

Comment:
I am seeing almost identical copy-pasted classes, and still don't see the benefit in term of making things simpler

Reply:
the purpose of this PR is not to make this stuff simpler, it is to cling to the spec and allowing to eliminate a few systematic errors this way. E.g., if the spec says that a protocol ought to be referenced, we can now expect a `Protocol` class instead of a `DiagLayer`. Also, attributes which are not present in the specification do no longer exist in our classes (e.g., parent refs in ECU-SHARED-DATA)

Reply:
ODX model is for serialization, for actual diagnostic encoding/decoding, you may want to take a look at the standard D-PDU API model.

Reply:
I use this ODX data to talk to ECUs all the time. as far as I know, the PDX files are even canonical source of diagnostics data uppon which all tools except candela studio build uppon...